README.rst
This is a fork of https://github.com/mnaberez/py65 since the author blocked me
when I tried to submit a pull request. Please direct any bug reports or pull
requests to this repository and I will do my best to handle them in a polite
and respectful fashion.

Nice65
====

Nice65 provides tools for simulating hardware based on 6502-like
microprocessors.  It has the following goals:

- Focus on ease of use and modularity rather than performance.  Nice65 is
  written in the Python programming language for productivity, while
  similar programs are written in C for performance.

- Enable simulations to be created for systems where it might have
  otherwise not been practical, such as homebuilt computers.

- Rigorously unit test all of the components.  While the tools provided
  by Nice65 may not always be perfect, their behavior is verified through
  tests so unexpected results are minimized.

Installation
------------

Nice65 packages are `available <http://pypi.python.org/pypi/nice65>`_ on the
Python Package Index (PyPI).  You download them from there or you can
use ``pip`` to automatically install or upgrade Nice65::

    $ pip install -U nice65

Devices
-------

The following devices are simulated at this time:

- ``mpu6502`` simulates the original NMOS 6502 microprocessor from MOS
  Technology, later known as Commodore Semiconductor Group (CSG). At this
  time, all of the documented opcodes are supported.  Support for the
  illegal opcodes is planned for the future.

- ``mpu65c02`` simulates a generic CMOS 65C02 microprocessor. There were
  several 65C02 versions from various manufacturers, some with more
  opcodes than others. This simulation is based on the W65C02S from the
  Western Design Center (WDC).

- ``mpu65org16`` simulates the 65Org16, a 6502-like microprocessor with a
  16-bit data bus and 32-bit address bus.  This microprocessor is a project
  of the `6502.org community <http://forum.6502.org/viewtopic.php?t=1824>`_
  and a `Verilog core <https://github.com/BigEd/verilog-6502/wiki>`_ for it
  has been implemented.

Monitor
-------

Nice65 includes a console-based machine language monitor (sometimes also called
a debugger).  This program, ``nice65mon``, allows you to interact with the
simulations that you build.  Its features include:

- Commands that are largely compatible with those used in the monitor of
  the popular VICE emulator for Commodore computers.

- Ability to load, dump, and fill memory.

- Simple assemble and disassemble capability, including support for labels
  and labels with offsets.

Documentation
-------------

Nice65 documentation is written using `Sphinx <http://sphinx.pocoo.org/>`_ and
is published to `http://nice65.readthedocs.org/
<http://nice65.readthedocs.org/>`_.


CHANGES.txt
2.0.0 (First nice65 fork)
-------------------------

- This is the first release of the nice65 fork, based on
  py65 `cdec5df`.

2.0.0.dev0 (Next Release)
-------------------------

 - Fixed a bug with character input that would cause characters to be
   dropped when pasting in larger amounts of text.  This makes it possible
   to paste programs into EhBASIC and Taliforth.  Patch by SamCoVT.

 - Support for some older Python versions has been dropped.  On Python 3,
   Py65 now requires Python 3.4 or later.  On Python 2, Py65 now requires
   Python 2.7.

 - added ``irq()`` and ``nmi()`` methods to the ``MPU`` class, so that
   interrupts can be simulated. Patch by Irmen de Jong.


1.1.0 (2018-07-01)
------------------

 - The ``Monitor`` class now allows the default memory to be supplied in
   the constructor.  Patch by Irmen de Jong.

 - Fixed a bug where setting the MPU via ``py65mon`` command line arguments
   would have no effect.  Reported by Michael A. Morris, patch by Ed Spittles.

 - The ``itoa()`` function in ``conversions.py`` now raises an error when an
   unsupported base is given.  Patch by Scot W. Stevenson.

 - The unused hexdump loader utility has been removed.

1.0.0 (2017-05-11)
------------------

 - Fixed a bug where the ordering of ``py65mon`` command line arguments
   produced different results.  Arguments can now be specified in any
   order.  Patch by Mario Keller.

 - Added new ``--input`` and ``--output`` arguments to ``py65mon`` that
   allow the addresses of the ``getc`` and ``putc`` handlers to be
   changed.  Patch by Mario Keller.

 - Fixed a 65C02 bug where the TSB and TRB instructions did not set
   the Z flag correctly.  Thanks to Kris Kennaway for reporting it.

0.24 (2015-03-31)
-----------------

 - Released as a universal wheel.

0.23 (2015-02-10)
-----------------

 - Added a workaround to $F001 output to catch encoding errors and
   display a "?" instead of crashing.  This condition can occur if
   the 6502 program sends bytes to $F001 that aren't compatible with
   the terminal's character encoding.

0.22 (2015-02-09)
-----------------

 - Fixed a bug where ``py65mon --rom`` would raise an exception
   on Python 3.

0.21 (2015-02-09)
-----------------

 - Added support for breakpoints in the monitor.  Contributed by
   Alessandro Gatti.

 - ASCII literals are now supported by the assembler.  The statement
   "LDA #'A'" is equivalent to "LDA #$41".

 - Fixed console input when run under Python 3 on Windows.  Closes #27.

0.20 (2014-05-08)
-----------------

  - Page wrapping for indexed indirect (X) operations on 65C02 has been
    restored.  This reverts the change introduced in 0.18.  We now believe
    that this mode works the same on the 65C02 as it does on the 6502.

0.19 (2014-03-12)
-----------------

  - Fixed 65C02 opcode $D2: CMP Zero Page, Indirect.

  - Blocking character input at $F005 has been removed.  The I/O area
    was designed to be compatible with Michal Kowalski's simulator,
    and it uses this address for another purpose.  Examples that depended
    on $F005 have been changed to use $F004.

0.18 (2014-01-30)
-----------------

  - Fixed a bug in RTS where popping $FFFF off the stack would cause
    the program counter to overflow to $10000.  It now wraps around
    to $0000 as it should.  Thanks to Gábor Lénárt for reporting it.

  - Fixed BRK on 65C02 so it clears the decimal flag.  On NMOS 6502, the
    decimal flag is unaffected on BRK.  On 65C02, it is always cleared.

  - Assembling now tolerates extra whitespace between opcode and operand.

  - Removed page wrap bug from JMP indirect on 65C02.

  - Removed page wrap bug from indexed indirect (X) operations on 65C02.

0.17 (2013-10-26)
-----------------

  - Added support for Python 3.2 and 3.3 based on work by David Beazley.

0.16 (2013-04-03)
-----------------

  - Fixed a bug in the monitor that caused loading from the command
    line with "--rom" to crash.

0.15 (2013-01-06)
-----------------

  - Disassembling can now wrap around memory if the start address
    given is greater than the end address.

  - Fixed the disassembler to accept a range of "start:end" instead of
    "start:end+1" for consistency with the other commands.

0.14 (2012-11-30)
-----------------

  - Assembling now detects syntax errors, overflows, and bad labels
    separately and shows specific error messages.

  - Fixed assembling 65C02 opcodes whose mnemonics have a digit
    such as "RMB3".

  - Reformatted source code to comply with PEP8.

  - Fixed a bug where the MPU status display would wrap unexpectedly
    on some terminals.

  - Added support for 65C02 opcode 0x89: BIT #.

  - Added support for 65C02 opcode 0x7C: JMP (abs,x).

  - Assembling now shows an Overflow Error if the statement won't
    fit at the top of memory.

  - Fixed a bug in the disassembler where instructions that wrap past
    the top of memory would not be displayed properly.

0.13 (2012-11-15)
-----------------

  - Fixed a bug where negative numbers could be entered
    for addresses in the monitor.

0.12 (2012-02-16)
-----------------

  - Fixed a bug that caused ``help cd`` to raise an exception
    in the monitor.

  - Fixed a bug in the 65C02 simulation where the opcode 0x7A
    was named "PHY" instead of "PLY", causing incorrect assembly
    and disassembly.  Thanks to Brian Cassidy for reporting it.

  - Fixed the cycle count of 0xD2 (CMP zero page indirect) in
    the 65C02 simulation.  Thanks to Brian Cassidy for reporting it.

  - Added "h" as a monitor shortcut for "help".

0.11 (2012-01-07)
-----------------

  - Added a new 65Org16 MPU simulation written by Ed Spittles.

  - The monitor now accepts command line arguments.  See
    ``py65mon --help`` for usage.  Contributed by Ed Spittles.

  - The monitor's load command can now fetch URLs.

  - Python versions earlier than 2.6 are no longer supported.

0.10 (2011-08-27)
-----------------


examples/65Org16.boot.asm
 .LIST

; bootrom for nice65 monitor in 65Org16 mode
; based on
; Intel Hex format loader by Ross Archer (9 February 2001, freeware)
; from http: http://www.6502.org/source/monitors/intelhex/intelhex.htm
;

; use this monitor like this:
;   nice65mon -m 65org16
;   load bootrom.bin fe00
;   goto fe00

START = $FFFFFe00

; I/O is memory-mapped in nice65:
PUTC     = $f001
GETC     = $f004

; Note that Hex format for 65Org16 uses ';' not ':' as the start of record mark
; also note that some fields are now composed of 16-bit elements:
; previously:
;  length offset type   data     checksum
;     :/08/E008/00/08090A0B0C0D0E0F/xx
; now
;     ;/10/E008/00/00080009000A000B000C000D000E000F/xx

; Zero-page storage
DPL       =  $00 ; data pointer (two bytes) used by PUTSTRI
DPH       =  $01 ; high of data pointer
RECLEN    =  $02 ; record length in bytes
START_LO  =  $03
START_HI  =  $04
RECTYPE   =  $05
CHKSUM    =  $06 ; record checksum accumulator
DLFAIL    =  $07 ; flag for download failure
TEMP      =  $08 ; save hex value
TMPHEX    =  $09 ; save another hex value

	; where the RAM program MUST have its first instruction
ENTRY_POINT  =  $0200

 .ORG START

        sei                     ; disable interrupts
        cld                     ; binary mode arithmetic
        ldx     #$1FFFF         ; Set up the stack pointer
        txs                     ;       "

        ; Download Intel hex.  The program you download MUST have its entry
        ; instruction (even if only a jump to somewhere else) at ENTRY_POINT.
HEXDNLD lda     #0
        sta     START_HI        ; store all programs in bank 0 (page 0) for now
        sta     DLFAIL          ;Start by assuming no D/L failure
        jsr     PUTSTRI
        .byte   13,10,13,10
        .byte   "Send 65Org16 code in"
        .byte   " variant Intel Hex format"
        .byte  " at 19200,n,8,1 ->"
        .byte   13,10
	.byte	0		; Null-terminate unless you prefer to crash.
HDWRECS jsr     GETSER          ; Wait for start of record mark ';'
        cmp     #';'
        bne     HDWRECS         ; not found yet
        ; Start of record marker has been found
        lda     #0
        sta     CHKSUM
        jsr     GETHEX          ; Get the record length
        sta     RECLEN          ; save it
        jsr     GET4HX          ; Get the 16-bit offset
        sta     START_LO
        jsr     GETHEX          ; Get the record type
        sta     RECTYPE         ; & save it
        bne     HDER1           ; end-of-record
        ldx     RECLEN          ; number of data bytes to write to memory
        ldy     #0              ; start offset at 0
HDLP1   jsr     GET4HX          ; Get the first/next/last data word
        sta     (START_LO),y    ; Save it to RAM
        iny                     ; update data pointer
        dex                     ; decrement character count
        dex                     ; ... twice
        bne     HDLP1
        jsr     GETHEX          ; get the checksum
        lda     CHKSUM
        bne     HDDLF1          ; If failed, report it
        ; Another successful record has been processed
        lda     #'#'            ; Character indicating record OK = '#'
        sta     PUTC            ; write it out but don't wait for output
        jmp     HDWRECS         ; get next record
HDDLF1  lda     #'F'            ; Character indicating record failure = 'F'
        sta     DLFAIL          ; download failed if non-zero
        sta     PUTC            ; write it to transmit buffer register
        jmp     HDWRECS         ; wait for next record start
HDER1   cmp     #1              ; Check for end-of-record type
        beq     HDER2
        jsr     PUTSTRI         ; Warn user of unknown record type
        .byte   13,10,13,10
        .byte   "Unknown record type $"
	.byte	0		; null-terminate unless you prefer to crash!
        lda     RECTYPE         ; Get it
	sta	DLFAIL		; non-zero --> download has failed
        jsr     PUTHEX          ; print it
	lda     #13		; but we'll let it finish so as not to
        jsr     PUTSER		; falsely start a new d/l from existing
        lda     #10		; file that may still be coming in for
        jsr     PUTSER		; quite some time yet.
	jmp	HDWRECS
	; We've reached the end-of-record record
HDER2   jsr     GETHEX          ; get the checksum
        lda     CHKSUM          ; Add previous checksum accumulator value
        beq     HDER3           ; checksum = 0 means we're OK!
        jsr     PUTSTRI         ; Warn user of bad checksum
        .byte   13,10,13,10
        .byte   "Bad record checksum!",13,10
        .byte   0		; Null-terminate or 6502 go bye-bye
        jmp     START
HDER3   lda     DLFAIL
        beq     HDEROK
        ;A download failure has occurred
        jsr     PUTSTRI
        .byte   13,10,13,10
        .byte   "Download Failed",13,10
        .byte   "Aborting!",13,10
	.byte	0		; null-terminate every string yada yada.
        jmp     START
HDEROK  jsr     PUTSTRI
        .byte   13,10,13,10
        .byte   "Download Successful!",13,10
        .byte   "Jumping to location $"
	.byte	0			; by now, I figure you know what this is for. :)
        lda	#HI(ENTRY_POINT)	; Print the entry point in hex
        jsr	PUTHEX
        lda	#LO(ENTRY_POINT)
	jsr	PUTHEX
        jsr	PUTSTRI
        .byte   13,10
        .byte   0		; stop lemming-like march of the program ctr. thru data
        jmp     ENTRY_POINT	; jump to canonical entry point

; For nice65, the input routine will block until a character arrives
GETSER  lda     GETC
        beq     GETSER
        rts

; get four ascii chars, adding both octets into the checksum
GET4HX  jsr     GETHEX
        asl a
        asl a
        asl a
        asl a
        asl a
        asl a
        asl a
        asl a
        sta     TMPHEX
        jsr     GETHEX
        ora     TMPHEX
        rts

; get two ascii chars, add into the checksum
GETHEX  jsr     GETSER
        jsr     MKNIBL  	; Convert to 0..F numeric
        asl     a
        asl     a
        asl     a
        asl     a       	; This is the upper nibble
        and     #$F0
        sta     TEMP
        jsr     GETSER
        jsr     MKNIBL
        ora     TEMP
        sta     TEMP
        clc
        adc     CHKSUM          ; Add in the checksum
        and     #$ff
        sta     CHKSUM          ;
        lda     TEMP
        rts             	; return with the nibble received

; Convert the ASCII nibble to numeric value from 0-F:
MKNIBL  cmp     #'9'+1  	; See if it's 0-9 or 'A'..'F' (no lowercase yet)
        bcc     MKNNH   	; If we borrowed, we lost the carry so 0..9
        sbc     #7+1    	; Subtract off extra 7 (sbc subtracts off one less)
        ; If we fall through, carry is set unlike direct entry at MKNNH
MKNNH   sbc     #'0'-1  	; subtract off '0' (if carry clear coming in)
        and     #$0F    	; no upper nibble no matter what
        rts             	; and return the nibble

; Put byte in A as hexydecascii
PUTHEX  pha             	;
        lsr a
        lsr a
        lsr a
        lsr a
        jsr     PRNIBL
        pla
PRNIBL  and     #$0F    	; strip off the low nibble
        cmp     #$0A
        bcc     NOTHEX  	; if it's 0-9, add '0' else also add 7
        adc     #6      	; Add 7 (6+carry=1), result will be carry clear
NOTHEX  adc     #'0'    	; If carry clear, we're 0-9
; Write the character in A as ASCII:
PUTSER  sta     PUTC
        rts

;Put the string following in-line until a NULL out to the console
PUTSTRI pla			; Get the low part of "return" address (data start address)
        sta     DPL
        pla
        sta     DPH             ; Get the high part of "return" address
                                ; (data start address)
        ; Note: actually we're pointing one short
PSINB   ldy     #1
        lda     (DPL),y         ; Get the next string character
        inc     DPL             ; update the pointer
        bne     PSICHO          ; if not, we're pointing to next character
        inc     DPH             ; account for page crossing
PSICHO  ora     #0              ; Set flags according to contents of Accumulator
        beq     PSIX1           ; don't print the final NULL
        jsr     PUTSER          ; write it out
        jmp     PSINB           ; back around
PSIX1   inc     DPL             ;
        bne     PSIX2           ;
        inc     DPH             ; account for page crossing
PSIX2   jmp     (DPL)           ; return to byte following final NULL
;
; Dummy interrupt handlers
GOIRQ
GONMI   RTI

; vectors
 .ORG $FFFFFFFA
NMIENT  .word     GONMI
RSTENT  .word     START
IRQENT  .word     GOIRQ
.end				; finally.  das Ende.
